src/installation/guides/zfs.md

```diff
@@ -0,0 +1,330 @@
+
+```
+# zfs create -o mountpoint=none zroot/ROOT
+# zfs create -o mountpoint=/ zroot/ROOT/void
+cannot mount '/': directory is not empty
+property may be set but unable to remount filesystem
+```
+
+## Base Installation
+
+To mount the system for installation export then re-import the pool at an
+alternate mountpoint:
+
+```
+# zpool export zroot
+# zpool import -R /mnt/void zroot
+```
+
+Populate the new installation with the following directories, which will provide
+mountpointsto the current system's devices:
+
+```
+# mkdir /mnt/void/{dev,sys,proc}
+```
+
+Then mount the directories required for a successfull installation:
+
+```
+# mount -B /dev /mnt/void/dev
+# mount -t devpts pts /mnt/void/dev/pts
+# mount -t proc proc /mnt/void/proc
+# mount -t sysfs sys /mnt/void/sys
+```
+
+Install the base system onto the drive along with any other packages needed to
+assist in configuration: (ie. NetworkManager, Vim, Nano, etc.)
+
+`# xbps-install -S -R https://alpha.de.repo.voidlinux.org/current -r /mnt/void
+base-system grub zfs`
+
+## GRUB Installation
+
+[Chroot(1)](https://man.voidlinux.org/chroot) into the new installation,
+invoking BASH to replace DASH as a more feature complete terminal, and
+specifying the terminal emulator to avoid errant behavior:
+
+`# TERM=linux chroot /mnt/void bash`
+
+Set the root password using [passwd(1)](https://man.voidlinux.org/passwd), which
+will prevent the user from being locked out at restart:
+
+`(chroot)# passwd root`
+
+Confirm the ZFS modules are loaded in the new system using
+[lsmod(8)](https://man.voidlinux.org/lsmod):
+
+```
+(chroot)$ lsmod | grep zfs
+zfs                  4022272  1
+zunicode              335872  1 zfs
+zlua                  172032  1 zfs
+zavl                   16384  1 zfs
+icp                   294912  1 zfs
+zcommon                86016  2 zfs,icp
+znvpair                69632  2 zfs,zcommon
+spl                   102400  5 zfs,icp,znvpair,zcommon,zavl
+```
+
+Set the new machine's hostname, replacing zfshost with the name of the new
+machine:
+
+`(chroot)# echo zfshost >/etc/hostname`
+
+Setup the ZFS cachefile, to ensure the new machine recognizes the created pool.
+
+`(chroot)# zpool set cachefile=/etc/zfs/zpool.cache zroot`
+
+Then notate the bootable system for GRUB's autoconfig:
+
+`(chroot)# zpool set bootfs=zroot/ROOT/void zroot`
+
+Also ensure GRUB recognizes the ZFS module.
+
+> **Note:**
+> 
+> When using id's set the ZPOOL_VDEV_NAME_PATH variable to 1.
+
+```
+(chroot)# ZPOOL_VDEV_NAME_PATH=1 grub-probe /
+zfs
+```
+
+> **Optional:**
+> 
+
+Install GRUB onto the drive specifying the path to the drive:
+
+```
+(chroot)# ZPOOL_VDEV_NAME_PATH=1 grub-install /dev/sda
+Installing for i386-pc platform.
+Installation finished. No error reported.
+```
+
+Create and modify `/etc/dracut.conf.d/zol.conf` to configure Dracut to include
+at least the following values:
+
+```
+hostonly="yes"
+nofsck="yes"
+add_dracutmodules+=" zfs "
```

Comment:
Can modern grub find `/boot` on zfs? Last I checked you needed an older pool version for this to work, have they fixed that?

Reply:
> Can modern grub find `/boot` on zfs? Last I checked you needed an older pool version for this to work, have they fixed that?

Any modern feature flags on your zpool will cause it to be no longer readable by GRUB. Unless you use other tools, `/boot` on ZFS with GRUB isn't advisable.

Reply:
I feel like this is a big footgun, Probably best to advise a separate partition for `/boot` or `/boot/efi` (or zfsbootmenu together with that).

Reply:
@zdykstra can you mention which features interfere with GRUB? I have `/boot` on zfs and all features are listed as 'active' or 'enabled'.

Reply:
It's surprisingly hard to find a definitive list of "these features will break GRUB". Instead, I'm able to find "these features are compatible with GRUB".  The most concise description I can find is the official OpenZFS guide to installing Ubuntu to ZFS - https://openzfs.github.io/openzfs-docs/Getting%20Started/Ubuntu/Ubuntu%2020.04%20Root%20on%20ZFS.html#step-2-disk-formatting .

As I understand it, if you have `feature@encryption` enabled on the pool, GRUB will refuse to read the pool, even if you don't actually have any encrypted filesystems. I haven't used GRUB to boot ZFS in quite a while, so I can't conclusively say beyond this though. I've moved on from GRUB and I use https://github.com/zdykstra/zfsbootmenu now to boot.

Reply:
@zdykstra On my local install I have:

    dozer  feature@encryption             enabled                        local

With not only `/` but also `/boot` on ZFS, with both sitting on a LUKS-encrypted partition, and I can boot with GRUB without issue, so simply having the feature enabled doesn't seem to block GRUB.

Now, I haven't tried to see what happens if I actually try to use native zfs encryption. I certainly imagine a natively-encrypted ZFS `/boot` GRUB would not be happy about.

(I think your zfsbootmenu looks like a better option in general, though I haven't had time to try to test using a working setup for non-UEFI boot.)

Reply:
You for sure can't use native encryption for `/boot` with GRUB, hah. More importantly, If there's an encrypted filesystem anywhere on the pool that contains `/boot`, GRUB will error out. Project Trident has a legacy/GRUB mode in their installer, and they were forced to skip creating an encrypted `/home/$user` filesystem on the same pool that `/boot` is on, even though GRUB has no need to touch `/home/$user`. Once `feature@encryption` is marked as `active`, things get hinky.

I think if anything is to be changed here, it should be to follow in the official OpenZFS footsteps and provide an explicit list of pool features to be enabled, or to split `/boot` into a new dedicated pool with only those known-good features enabled. 

Not to derail, but `zfsbootmenu` works well under a legacy/BIOS setup by utilizing `syslinux` as the first stage bootloader.

Reply:
So perhaps it's not about 'enabled' features, but 'active' features. (Since I'm on the latest zfs and have all features enabled.)

I think providing an explicit list of features to be enabled would be greatly preferable. There's advantages to having `/boot` as part of the same pool (e.g. you don't need to worry about `/boot` running out of space).

(Is [this](https://github.com/zdykstra/zfsbootmenu/tree/syslinux-docs#syslinuxextlinux) still the best resource for setting up `zfsbootmenu` for a non-EFI system?)

Reply:
I have 26 features enabled on my current pool, there are only 12 features enabled on the boot pool in the OpenZFS documentation. Comparing those two, we get this:

```
diff -u boot-features-sorted all-features-sorted 
--- boot-features-sorted	2020-06-25 11:02:49.915090112 -0500
+++ all-features-sorted	2020-06-25 11:02:41.771950496 -0500
@@ -1,12 +1,26 @@
+feature@allocation_classes
 feature@async_destroy
 feature@bookmarks
+feature@bookmark_v2
+feature@device_removal
+feature@edonr
 feature@embedded_data
 feature@empty_bpobj
 feature@enabled_txg
+feature@encryption
 feature@extensible_dataset
 feature@filesystem_limits
 feature@hole_birth
 feature@large_blocks
+feature@large_dnode
 feature@lz4_compress
+feature@multi_vdev_crash_dump
+feature@obsolete_counts
+feature@project_quota
+feature@resilver_defer
+feature@sha512
+feature@skein
 feature@spacemap_histogram
+feature@spacemap_v2
+feature@userobj_accounting
 feature@zpool_checkpoint
```

I don't have a nuanced enough understanding of what each of those features enables, but they're specifically skipped when creating `bpool` in the Ubuntu install guide. I think at the minimum, if the OpenZFS development themselves follows the practice of limiting which feature flags are turned on, we should as well.

Reply:
Do you think it's feasible to maintain a list like that?

Reply:
> Do you think it's feasible to maintain a list like that?

I don't see any real issue with it. It doesn't change when OpenZFS adds features, it changes with GRUB adds support for new features. Since they haven't added any new feature support since 2015, this seems to be relatively set in stone.

Reply:
The follow-up then is: is it worth recommending GRUB + ZFS when it has so many limitations?